File: submissions/EXAMPLE.username/bot.py
"""
BotPit minimal-but-safe Python bot — HMAC code-bot path.

Spec: https://www.botpit.io/llms.txt
"""

from __future__ import annotations

import os
import sys
import time
import enum
import json
import hmac
import hashlib
import logging
from dataclasses import dataclass
from typing import Any, Dict, Optional

import requests

API_BASE = os.getenv("BOTPIT_API_BASE", "https://www.botpit.io")
PUBKEY = os.environ.get("BOTPIT_AGENT_PUBKEY")
SECRET = os.environ.get("BOTPIT_AGENT_SECRET")
PAIR = os.getenv("BOTPIT_PAIR", "BTC-USDT")
TICK_SECONDS = int(os.getenv("BOTPIT_TICK_SECONDS", "10"))

# Note: env-var validation is deferred to run() so the module can be
# imported without credentials (e.g. by CI validators or unit tests).
# The bot only fails fast when you actually try to start it.

logging.basicConfig(format="[%(asctime)s] %(message)s", datefmt="%H:%M:%S", level=logging.INFO)
log = logging.getLogger("bot")


class Action(enum.Enum):
    HOLD = "hold"
    OPEN_LONG = "open_long"
    OPEN_SHORT = "open_short"
    CLOSE = "close"


@dataclass
class Decision:
    action: Action
    size_pct: float = 10.0
    leverage: int = 5
    stop_pct: float = 1.5
    take_profit_pct: float = 3.0


class BotPit:
    def __init__(self, base: str, pubkey: str, secret: str):
        self.base = base.rstrip("/")
        self.pubkey = pubkey
        self.secret = secret.encode("utf-8")
        self.session = requests.Session()

    def _sign(self, body: str) -> Dict[str, str]:
        t = int(time.time() * 1000)
        mac = hmac.new(self.secret, f"{t}.{body}".encode("utf-8"), hashlib.sha256)
        return {
            "Agent-Arena-Key": self.pubkey,
            "Agent-Arena-Signature": f"t={t},v1={mac.hexdigest()}",
        }

    def tournament(self) -> dict:
        r = self.session.get(f"{self.base}/api/v1/tv/tournament",
                             headers={**self._sign(""), "Content-Type": "application/json"}, timeout=8)
        r.raise_for_status()
        return r.json()

    def state(self) -> dict:
        r = self.session.get(f"{self.base}/api/v1/tv/state",
                             headers={**self._sign(""), "Content-Type": "application/json"}, timeout=8)
        r.raise_for_status()
        return r.json()

    def send_signal(self, *, side: str, pair: str, size_pct: float, leverage: int,
                    nonce: Optional[int] = None) -> dict:
        body_dict = {
            "nonce": nonce if nonce is not None else int(time.time() * 1000),
            "pair": pair, "side": side, "order_type": "market",
            "size": {"mode": "pct_equity", "value": size_pct}, "leverage": leverage,
        }
        body = json.dumps(body_dict, separators=(",", ":"))
        r = self.session.post(f"{self.base}/api/v1/signals", data=body,
                              headers={**self._sign(body), "Content-Type": "application/json"}, timeout=8)
        if r.status_code >= 400:
            log.warning("signal rejected: %s %s", r.status_code, r.text[:240])
            return {"status": "rejected", "http": r.status_code, "raw": r.text}
        return r.json()

    def log_decision(self, *, action: str, reason: str,
                     payload: Optional[Dict[str, Any]] = None) -> None:
        """Append a decision-trace record to /api/v1/decisions. Lets dashboards
        render WHY the bot chose what it chose without recomputing the strategy.
        Fire-and-forget — failures are logged but don't break the tick.
        Platform caps to last 100 rows per agent automatically."""
        body_dict: Dict[str, Any] = {"action": action, "reason": reason}
        if payload is not None:
            body_dict["payload"] = payload
        body = json.dumps(body_dict, separators=(",", ":"))
        try:
            r = self.session.post(f"{self.base}/api/v1/decisions", data=body,
                                  headers={**self._sign(body), "Content-Type": "application/json"}, timeout=4)
            if r.status_code >= 400:
                log.warning("decision rejected: %s %s", r.status_code, r.text[:200])
        except requests.RequestException as e:
            log.warning("decision log failed: %s", e)


BINANCE_MARK = "https://fapi.binance.com/fapi/v1/premiumIndex"

def get_mark_price(pair: str) -> float:
    symbol = pair.replace("-", "")
    r = requests.get(BINANCE_MARK, params={"symbol": symbol}, timeout=5)
    r.raise_for_status()
    return float(r.json()["markPrice"])


@dataclass
class Snapshot:
    equity_usd: float
    return_pct: float
    drawdown_pct: float
    open_position: Optional[Dict[str, Any]]
    last_fill_price: Optional[float]
    pair_config: Dict[str, Any]


@dataclass
class StopMemory:
    stop_price: Optional[float] = None
    take_profit_price: Optional[float] = None


memory = StopMemory()


def decide(snap: Snapshot, mark_price: float) -> Decision:
    """PLACEHOLDER STRATEGY — replace with your own.

    Examples to ask your LLM for:
      - "Open long when mark price drops 1% from the 60-tick rolling high.
         Close when up 0.5% or down 1%."
      - "Buy when RSI(14) crosses below 30, close when it crosses above 70."

    Return Decision(Action.HOLD) | Decision(Action.OPEN_LONG, ...) |
    Decision(Action.OPEN_SHORT, ...) | Decision(Action.CLOSE).
    """
    return Decision(Action.HOLD)


def watch_stops(snap: Snapshot, mark_price: float) -> Optional[Decision]:
    if snap.open_position is None:
        return None
    side = snap.open_position["side"]
    if memory.stop_price is not None:
        if side == "long" and mark_price <= memory.stop_price:
            log.info("STOP HIT (long) — mark %.2f <= stop %.2f", mark_price, memory.stop_price)
            return Decision(Action.CLOSE)
        if side == "short" and mark_price >= memory.stop_price:
            log.info("STOP HIT (short) — mark %.2f >= stop %.2f", mark_price, memory.stop_price)
            return Decision(Action.CLOSE)
    if memory.take_profit_price is not None:
        if side == "long" and mark_price >= memory.take_profit_price:
            log.info("TP HIT (long) — mark %.2f >= tp %.2f", mark_price, memory.take_profit_price)
            return Decision(Action.CLOSE)
        if side == "short" and mark_price <= memory.take_profit_price:
            log.info("TP HIT (short) — mark %.2f <= tp %.2f", mark_price, memory.take_profit_price)
            return Decision(Action.CLOSE)
    return None


def build_snapshot(api: BotPit, pair: str, rules: dict) -> Snapshot:
    s = api.state()
    open_pos = next((p for p in s["positions"] if p["pair"] == pair), None)
    last_fill_price = None
    if open_pos:
        for f in s["recent_fills"]:
            if f["pair"] == pair and f["side"] == open_pos["side"]:
                last_fill_price = f["price"]
                break
    return Snapshot(
        equity_usd=s["equity"]["current_usd"],
        return_pct=s["equity"]["return_pct"],
        drawdown_pct=s["equity"]["drawdown_pct"],
        open_position=open_pos,
        last_fill_price=last_fill_price,
        pair_config=rules,
    )


def apply_decision(api: BotPit, decision: Decision, snap: Snapshot, pair: str) -> None:
    if decision.action == Action.HOLD:
        return
    if decision.action in (Action.OPEN_LONG, Action.OPEN_SHORT):
        if snap.open_position is not None:
            log.info("decide() wants %s but already %s; close first.",
                     decision.action.value, snap.open_position["side"])
            return
        side = "long" if decision.action == Action.OPEN_LONG else "short"
        resp = api.send_signal(side=side, pair=pair, size_pct=decision.size_pct, leverage=decision.leverage)
        log.info("OPEN %s sent -> %s", decision.action.value, resp.get("status"))
        try:
            mark = get_mark_price(pair)
            if decision.action == Action.OPEN_LONG:
                memory.stop_price = mark * (1 - decision.stop_pct / 100)
                memory.take_profit_price = mark * (1 + decision.take_profit_pct / 100)
            else:
                memory.stop_price = mark * (1 + decision.stop_pct / 100)
                memory.take_profit_price = mark * (1 - decision.take_profit_pct / 100)
            log.info("stop set @ %.2f, tp @ %.2f", memory.stop_price, memory.take_profit_price)
        except Exception as e:
            log.warning("couldn't set client-side stop: %s", e)
        return
    if decision.action == Action.CLOSE:
        if snap.open_position is None:
            return
        resp = api.send_signal(side="close", pair=pair, size_pct=100, leverage=1)
        log.info("CLOSE sent -> %s", resp.get("status"))
        memory.stop_price = None
        memory.take_profit_price = None


def run() -> None:
    if not PUBKEY or not SECRET:
        sys.exit(
            "BOTPIT_AGENT_PUBKEY / BOTPIT_AGENT_SECRET not set. "
            "Copy .env.example to .env and paste the keypair from "
            "https://www.botpit.io/agents/<your-agent-id>."
        )
    api = BotPit(API_BASE, PUBKEY, SECRET)
    log.info("starting up against %s", API_BASE)
    t = api.tournament()
    rules, tournament = t["rules"], t["tournament"]
    log.info("tournament: %s (ends %s)", tournament["name"], tournament["ends_at"])
    log.info("rules: leverage_cap=%dx allowed_pairs=%s starting_equity=$%s",
             rules["leverage_cap"], rules["allowed_pairs"], rules["starting_equity_usd"])
    if PAIR not in rules["allowed_pairs"]:
        sys.exit(f"PAIR={PAIR} not in allowed_pairs {rules['allowed_pairs']}; set BOTPIT_PAIR.")

    last_heartbeat = 0.0
    HEARTBEAT_EVERY = 60

    while True:
        try:
            snap = build_snapshot(api, PAIR, rules)
            mark = get_mark_price(PAIR)
            stop_decision = watch_stops(snap, mark)
            if stop_decision:
                apply_decision(api, stop_decision, snap, PAIR)
                snap = build_snapshot(api, PAIR, rules)
            decision = decide(snap, mark)
            apply_decision(api, decision, snap, PAIR)
            now = time.time()
            if now - last_heartbeat > HEARTBEAT_EVERY:
                pos_str = (
                    f"{snap.open_position['side']} {snap.open_position['size_units']:.4f} {PAIR} @ ${snap.open_position['entry_price']:.2f}"
                    if snap.open_position else "flat"
                )
                log.info("hb: equity=$%.2f return=%+.2f%% dd=%.2f%% pos=[%s] mark=%.2f",
                         snap.equity_usd, snap.return_pct, snap.drawdown_pct, pos_str, mark)
                last_heartbeat = now
        except requests.RequestException as e:
            log.warning("network blip: %s -- backing off 5s", e)
            time.sleep(5)
            continue
        except Exception as e:
            log.exception("unexpected error: %s", e)
            time.sleep(5)
            continue
        time.sleep(TICK_SECONDS)


if __name__ == "__main__":
    run()

File: code-bot/typescript/bot.ts
/**
 * BotPit minimal-but-safe TypeScript bot — HMAC code-bot path.
 *
 * Spec: https://www.botpit.io/llms.txt
 *
 * Architecture: every TICK_SECONDS we
 *   1. Read /api/v1/tv/state              — what's actually true
 *   2. Run watchStops()                   — fire client-side stops if hit
 *   3. Run decide()                       — your strategy decides next action
 *   4. Apply the decision (sendSignal)    — POST /api/v1/signals (HMAC-signed)
 *   5. Log a heartbeat                    — so you can verify uptime
 *
 * Where your strategy goes: the `decide()` function below. Everything
 * else is plumbing.
 */

import crypto from "node:crypto";

// ---------- Config ----------

const API_BASE = process.env.BOTPIT_API_BASE ?? "https://www.botpit.io";
const PUBKEY = process.env.BOTPIT_AGENT_PUBKEY;
const SECRET = process.env.BOTPIT_AGENT_SECRET;
const PAIR = process.env.BOTPIT_PAIR ?? "BTC-USDT";
const TICK_SECONDS = parseInt(process.env.BOTPIT_TICK_SECONDS ?? "10", 10);

// Note: env-var validation is deferred to run() so the module can be
// imported without credentials (e.g. by CI validators or unit tests).
// The bot only fails fast when you actually try to start it.

// ---------- Strategy interface ----------

type Decision =
  | { action: "hold" }
  | { action: "close" }
  | {
      action: "open_long" | "open_short";
      sizePct: number;
      leverage: number;
      stopPct: number;
      takeProfitPct?: number;
    };

// ---------- API types ----------

type Position = {
  pair: string;
  side: "long" | "short";
  size_units: number;
  entry_price: number;
  leverage: number;
  unrealized_pnl_usd: number;
  opened_at: string;
};

type Fill = {
  pair: string;
  side: "long" | "short";
  price: number;
  size_units: number;
  size_usd: number;
  fee_usd: number;
  slippage_bps: number;
  filled_at: string;
};

type StateResponse = {
  tournament: { id: string; name: string | null; kind: string; ends_at: string };
  equity: {
    starting_usd: number; current_usd: number; peak_usd: number;
    return_pct: number; drawdown_pct: number;
    realized_pnl_usd: number; unrealized_pnl_usd: number; as_of: string | null;
  };
  positions: Position[];
  recent_fills: Fill[];
  recent_signals: Array<{
    signal_id: string; nonce: number; status: string;
    reason_code: string | null; reason: string | null;
    t_received: string; t_processed: string | null;
  }>;
};

type TournamentResponse = {
  tournament: {
    id: string; name: string; kind: string; state: string;
    starts_at: string; ends_at: string;
    league: { name: string; tier: number };
  };
  rules: {
    starting_equity_usd: number; leverage_cap: number;
    allowed_pairs: string[];
    fee_bps_taker: number; fee_bps_maker: number; slippage_bps_market: number;
    dq_threshold_pct: number;
  };
  scoring: { formula: string; drawdown_penalty_k: number };
};

// ---------- HMAC signing + API client ----------

function sign(body: string): { "Agent-Arena-Key": string; "Agent-Arena-Signature": string } {
  const t = Date.now();
  const mac = crypto.createHmac("sha256", SECRET!).update(`${t}.${body}`).digest("hex");
  return {
    "Agent-Arena-Key": PUBKEY!,
    "Agent-Arena-Signature": `t=${t},v1=${mac}`,
  };
}

async function apiGet<T>(path: string): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    headers: { ...sign(""), "Content-Type": "application/json" },
  });
  if (!res.ok) throw new Error(`${path}: ${res.status} ${await res.text()}`);
  return (await res.json()) as T;
}

async function sendSignal(
  side: "long" | "short" | "close",
  pair: string,
  sizePct: number,
  leverage: number
): Promise<{ status: string }> {
  const bodyObj = {
    nonce: Date.now(),
    pair,
    side,
    order_type: "market" as const,
    size: { mode: "pct_equity" as const, value: sizePct },
    leverage,
  };
  const body = JSON.stringify(bodyObj);
  const res = await fetch(`${API_BASE}/api/v1/signals`, {
    method: "POST",
    headers: { ...sign(body), "Content-Type": "application/json" },
    body,
  });
  if (res.status >= 400) {
    const text = await res.text();
    console.warn(`[bot] signal rejected: ${res.status} ${text.slice(0, 240)}`);
    return { status: "rejected" };
  }
  return (await res.json()) as { status: string };
}

// ---------- Mark price (Binance public futures) ----------

async function getMarkPrice(pair: string): Promise<number> {
  const symbol = pair.replace("-", "");
  const res = await fetch(`https://fapi.binance.com/fapi/v1/premiumIndex?symbol=${symbol}`);
  if (!res.ok) throw new Error(`mark price ${res.status}`);
  const j = (await res.json()) as { markPrice: string };
  return parseFloat(j.markPrice);
}

// ---------- Local stop memory ----------
const memory: { stopPrice: number | null; takeProfitPrice: number | null } = {
  stopPrice: null,
  takeProfitPrice: null,
};

// ---------- Strategy — REPLACE ME ----------

type Snapshot = {
  equityUsd: number;
  returnPct: number;
  drawdownPct: number;
  openPosition: Position | null;
  lastFillPrice: number | null;
  rules: TournamentResponse["rules"];
};

function decide(_snap: Snapshot, _markPrice: number): Decision {
  // PLACEHOLDER STRATEGY — replace with your own.
  //
  // Examples to ask your LLM for:
  //   - "Open long when mark price drops 1% from the 60-tick rolling high.
  //      Close when up 0.5% or down 1%."
  //   - "Buy when RSI(14) crosses below 30, close when it crosses above 70."
  return { action: "hold" };
}

// ---------- Stop watcher ----------

function watchStops(snap: Snapshot, markPrice: number): Decision | null {
  if (!snap.openPosition) return null;
  const side = snap.openPosition.side;
  if (memory.stopPrice !== null) {
    if (side === "long" && markPrice <= memory.stopPrice) {
      console.log(`[bot] STOP HIT (long) — mark ${markPrice} <= stop ${memory.stopPrice}`);
      return { action: "close" };
    }
    if (side === "short" && markPrice >= memory.stopPrice) {
      console.log(`[bot] STOP HIT (short) — mark ${markPrice} >= stop ${memory.stopPrice}`);
      return { action: "close" };
    }
  }
  if (memory.takeProfitPrice !== null) {
    if (side === "long" && markPrice >= memory.takeProfitPrice) {
      console.log(`[bot] TP HIT (long) — mark ${markPrice} >= tp ${memory.takeProfitPrice}`);
      return { action: "close" };
    }
    if (side === "short" && markPrice <= memory.takeProfitPrice) {
      console.log(`[bot] TP HIT (short) — mark ${markPrice} <= tp ${memory.takeProfitPrice}`);
      return { action: "close" };
    }
  }
  return null;
}

// ---------- Main loop ----------

async function buildSnapshot(rules: TournamentResponse["rules"]): Promise<Snapshot> {
  const s = await apiGet<StateResponse>("/api/v1/tv/state");
  const openPosition = s.positions.find((p) => p.pair === PAIR) ?? null;
  let lastFillPrice: number | null = null;
  if (openPosition) {
    for (const f of s.recent_fills) {
      if (f.pair === PAIR && f.side === openPosition.side) {
        lastFillPrice = f.price;
        break;
      }
    }
  }
  return {
    equityUsd: s.equity.current_usd,
    returnPct: s.equity.return_pct,
    drawdownPct: s.equity.drawdown_pct,
    openPosition,
    lastFillPrice,
    rules,
  };
}

async function applyDecision(decision: Decision, snap: Snapshot): Promise<void> {
  if (decision.action === "hold") return;

  if (decision.action === "open_long" || decision.action === "open_short") {
    if (snap.openPosition) {
      console.log(`[bot] decide() wants ${decision.action} but already ${snap.openPosition.side}; close first.`);
      return;
    }
    const side = decision.action === "open_long" ? "long" : "short";
    const resp = await sendSignal(side, PAIR, decision.sizePct, decision.leverage);
    console.log(`[bot] OPEN ${decision.action} sent -> ${resp.status}`);
    try {
      const mark = await getMarkPrice(PAIR);
      const stopPct = decision.stopPct;
      const tpPct = decision.takeProfitPct ?? stopPct * 2;
      if (decision.action === "open_long") {
        memory.stopPrice = mark * (1 - stopPct / 100);
        memory.takeProfitPrice = mark * (1 + tpPct / 100);
      } else {
        memory.stopPrice = mark * (1 + stopPct / 100);
        memory.takeProfitPrice = mark * (1 - tpPct / 100);
      }
      console.log(`[bot] stop set @ ${memory.stopPrice.toFixed(2)}, tp @ ${memory.takeProfitPrice.toFixed(2)}`);
    } catch (e) {
      console.warn(`[bot] couldn't set client-side stop: ${(e as Error).message}`);
    }
    return;
  }

  if (decision.action === "close") {
    if (!snap.openPosition) return;
    const resp = await sendSignal("close", PAIR, 100, 1);
    console.log(`[bot] CLOSE sent -> ${resp.status}`);
    memory.stopPrice = null;
    memory.takeProfitPrice = null;
  }
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

async function run(): Promise<void> {
  if (!PUBKEY || !SECRET) {
    console.error(
      "BOTPIT_AGENT_PUBKEY / BOTPIT_AGENT_SECRET not set. Copy .env.example to .env and paste the keypair from https://www.botpit.io/agents/<your-agent-id>."
    );
    process.exit(1);
  }
  console.log(`[bot] starting up against ${API_BASE}`);
  const t = await apiGet<TournamentResponse>("/api/v1/tv/tournament");
  const rules = t.rules;
  console.log(`[bot] tournament: ${t.tournament.name} (ends ${t.tournament.ends_at})`);
  console.log(
    `[bot] rules: leverage_cap=${rules.leverage_cap}x allowed_pairs=${rules.allowed_pairs.join(",")} starting_equity=$${rules.starting_equity_usd}`
  );
  if (!rules.allowed_pairs.includes(PAIR)) {
    console.error(`PAIR=${PAIR} not in allowed_pairs ${rules.allowed_pairs.join(",")}`);
    process.exit(1);
  }

  let lastHeartbeat = 0;
  const HEARTBEAT_EVERY_MS = 60_000;

  for (;;) {
    try {
      let snap = await buildSnapshot(rules);
      const mark = await getMarkPrice(PAIR);

      const stopDecision = watchStops(snap, mark);
      if (stopDecision) {
        await applyDecision(stopDecision, snap);
        snap = await buildSnapshot(rules);
      }

      const decision = decide(snap, mark);
      await applyDecision(decision, snap);

      const now = Date.now();
      if (now - lastHeartbeat > HEARTBEAT_EVERY_MS) {
        const posStr = snap.openPosition
          ? `${snap.openPosition.side} ${snap.openPosition.size_units.toFixed(4)} ${PAIR} @ $${snap.openPosition.entry_price.toFixed(2)}`
          : "flat";
        console.log(
          `[bot] hb: equity=$${snap.equityUsd.toFixed(2)} return=${snap.returnPct >= 0 ? "+" : ""}${snap.returnPct.toFixed(2)}% dd=${snap.drawdownPct.toFixed(2)}% pos=[${posStr}] mark=${mark.toFixed(2)}`
        );
        lastHeartbeat = now;
      }
    } catch (e) {
      console.warn(`[bot] tick failed: ${(e as Error).message} -- backing off 5s`);
      await sleep(5000);
      continue;
    }

    await sleep(TICK_SECONDS * 1000);
  }
}

run().catch((e) => {
  console.error(e);
  process.exit(1);
});
